### PR TITLE
[ft] Handle nullptr when destroying freetype font

### LINKED ISSUE #2493: NULL pointer access due to memory allocation failing...
While testing what happens when memory allocations fail using clang+ASAN I stumbled upon a crash in HarfBuzz 2.6.4, which still seems to be present in git HEAD.

The backtrace relevant for HarfBuzz for the crash is here, where it dereferences the NULL pointer argument passed to it:

```
    #0 0xd3019f in _hb_ft_font_destroy(void*) harfbuzz/src/hb-ft.cc:120:7
    #1 0xd2c525 in hb_font_set_funcs harfbuzz/src/hb-font.cc:1673:7
    #2 0xd30d8e in _hb_ft_font_set_funcs(hb_font_t*, FT_FaceRec_*, bool) harfbuzz/src/hb-ft.cc:559:3
    #3 0xd30c45 in hb_ft_font_create harfbuzz/src/hb-ft.cc:698:3
```

To get into this state is a bit complicated as `calloc()` needs to return `NULL` in multiple locations, the first being here:

```
#0  0x00000000010dffa7 in hb_object_create<hb_blob_t> () at harfbuzz/src/hb-object.hh:227
#1  0x00000000010dfca9 in hb_blob_create (data=0x384c475 "", length=86828, mode=HB_MEMORY_MODE_READONLY, user_data=0x61a000008488, destroy=0x0) at harfbuzz/src/hb-blob.cc:94
#2  0x0000000000d3053d in hb_ft_face_create (ft_face=0x61a000008488, destroy=0x0) at harfbuzz/src/hb-ft.cc:615
#3  0x0000000000d30c1d in hb_ft_font_create (ft_face=0x61a000008488, destroy=0x0) at harfbuzz/src/hb-ft.cc:695
```

Next `calloc()` returns `NULL` again here:

```
#0  0x00000000010e2e0e in _hb_face_for_data_closure_create (blob=0x17e46c0 <_hb_NullPool>, index=0) at harfbuzz/src/hb-face.cc:145
#1  0x00000000010e2ca1 in hb_face_create (blob=0x17e46c0 <_hb_NullPool>, index=0) at harfbuzz/src/hb-face.cc:205
#2  0x0000000000d30586 in hb_ft_face_create (ft_face=0x61a000008488, destroy=0x0) at harfbuzz/src/hb-ft.cc:619
#3  0x0000000000d30c1d in hb_ft_font_create (ft_face=0x61a000008488, destroy=0x0) at harfbuzz/src/hb-ft.cc:695
```

And then `calloc()` returns `NULL` here:

```
#0  0x0000000000d2f917 in hb_object_create<hb_font_t> () at harfbuzz/src/hb-object.hh:227
#1  0x0000000000d2ae78 in _hb_font_create (face=0x3982520 <_hb_Null_hb_face_t>) at harfbuzz/src/hb-font.cc:1349
#2  0x0000000000d2ae35 in hb_font_create (face=0x3982520 <_hb_Null_hb_face_t>) at harfbuzz/src/hb-font.cc:1376
#3  0x0000000000d30c2a in hb_ft_font_create (ft_face=0x61a000008488, destroy=0x0) at harfbuzz/src/hb-ft.cc:696
```

The fourth time `calloc()` returns `NULL` is here:

```
#0  0x0000000000d31b89 in _hb_ft_font_create (ft_face=0x61a000008488, symbol=false, unref=false) at harfbuzz/src/hb-ft.cc:89
#1  0x0000000000d30d75 in _hb_ft_font_set_funcs (font=0x39500e0 <_hb_Null_hb_font_t>, ft_face=0x61a000008488, unref=false) at harfbuzz/src/hb-ft.cc:561
#2  0x0000000000d30c46 in hb_ft_font_create (ft_face=0x61a000008488, destroy=0x0) at harfbuzz/src/hb-ft.cc:698
```

This last call implies that `hb_font_set_funcs()` is called with an Null font object, for which `hb_object_is_immuatble()` returns `true` causing the passed `destroy` callback `_hb_ft_font_destroy()` to be called with `NULL` which then results in the crash mentioned at the top of this issue:

```
#0  _hb_ft_font_destroy (data=0x0) at harfbuzz/src/hb-ft.cc:116
#1  0x0000000000d2c526 in hb_font_set_funcs (font=0x39500e0 <_hb_Null_hb_font_t>, klass=0x6140000c8a48, font_data=0x0, destroy=0xd30130 <_hb_ft_font_destroy(void*)>) at harfbuzz/src/hb-font.cc:1673
#2  0x0000000000d30d8f in _hb_ft_font_set_funcs (font=0x39500e0 <_hb_Null_hb_font_t>, ft_face=0x61a000008488, unref=false) at harfbuzz/src/hb-ft.cc:559
#3  0x0000000000d30c46 in hb_ft_font_create (ft_face=0x61a000008488, destroy=0x0) at harfbuzz/src/hb-ft.cc:698
```

Looking at the [source code](https://github.com/harfbuzz/harfbuzz/blob/master/src/hb-ft.cc#L119) it is clear that when `_hb_ft_font_destroy()` receives a `NULL` pointer, it will try to dereference it.

### PR DESCRIPTION
**Pull request by sebras**
Fixes #2493

### CODE REVIEW

**Comment by behdad** [comment]
I'll review. Thanks.

File: src/hb-ft.cc [hunk 1]
```diff
@@ -120,6 +120,9 @@ _hb_ft_font_destroy (void *data)
 {
   hb_ft_font_t *ft_font = (hb_ft_font_t *) data;
 
+  if (unlikely (ft_font == nullptr))
```
Comment (ebraminio) [inline src/hb-ft.cc:123]: we recently had `ft_font == nullptr` to `!ft_font` like cleanup

Comment (sebras) [reply src/hb-ft.cc]: That's an easy change. I'll fix.

**Comment by behdad** [comment]
Thanks. Let me review.

**Comment by ebraminio** [comment]
https://github.com/google/oss-fuzz/issues/1675#issuecomment-410371405

ulimit looked like that systematic solution but isn't, apparently we have to handle ourselves like https://github.com/open62541/open62541/pull/2188/files

**Comment by behdad** [comment]
Thank you for the full analysis. I've confirmed this now.

**Pull request merged**